aws-lc-rs/src/kem.rs

```diff
@@ -0,0 +1,530 @@
+        let mut ciphertext_len = self.algorithm.ciphertext_size();
+        let mut shared_secret_len = self.algorithm.shared_secret_size();
+        let mut ciphertext: Vec<u8> = vec![0u8; ciphertext_len];
+        let mut shared_secret: Vec<u8> = vec![0u8; shared_secret_len];
+
+        let ctx = LcPtr::new(unsafe { EVP_PKEY_CTX_new(*self.evp_pkey, null_mut()) })?;
+
+        if 1 != unsafe {
+            EVP_PKEY_encapsulate(
+                *ctx,
+                ciphertext.as_mut_ptr(),
+                &mut ciphertext_len,
+                shared_secret.as_mut_ptr(),
+                &mut shared_secret_len,
+            )
+        } {
+            return Err(Unspecified);
+        }
+
+        ciphertext.truncate(ciphertext_len);
+        shared_secret.truncate(shared_secret_len);
+
+        Ok((
+            Ciphertext::new(ciphertext),
+            SharedSecret::new(shared_secret.into_boxed_slice()),
+        ))
+    }
+
+    /// Creates a new KEM public key from raw bytes. This method is MUST NOT be used to generate
+    /// a new public key, rather it should be used to construct `PublicKey` that was previously serialized
+    /// to raw bytes.
+    ///
+    /// `alg` is the [`Algorithm`] to be associated with the generated `PublicKey`
+    ///
+    /// `bytes` is a slice of raw bytes representing a `PublicKey`
+    ///
+    /// # Errors
+    /// `error::KeyRejected` when operation fails during key creation.
+    pub fn new(alg: &'static Algorithm<Id>, bytes: &[u8]) -> Result<Self, KeyRejected> {
+        match bytes.len().cmp(&alg.public_key_size()) {
+            Ordering::Less => Err(KeyRejected::too_small()),
+            Ordering::Greater => Err(KeyRejected::too_large()),
+            Ordering::Equal => Ok(()),
+        }?;
+        let pubkey = LcPtr::new(unsafe {
+            EVP_PKEY_kem_new_raw_public_key(alg.id.nid(), bytes.as_ptr(), bytes.len())
+        })?;
+        Ok(EncapsulationKey {
+            algorithm: alg,
+            evp_pkey: pubkey,
+            pub_key: bytes.into(),
+        })
+    }
+}
+
+unsafe impl<Id> Send for EncapsulationKey<Id> where Id: AlgorithmIdentifier {}
+
+unsafe impl<Id> Sync for EncapsulationKey<Id> where Id: AlgorithmIdentifier {}
+
+impl<Id> Drop for EncapsulationKey<Id>
+where
+    Id: AlgorithmIdentifier,
+{
+    fn drop(&mut self) {
+        self.pub_key.zeroize();
+    }
+}
+
+impl<Id> AsRef<[u8]> for EncapsulationKey<Id>
+where
+    Id: AlgorithmIdentifier,
+{
+    fn as_ref(&self) -> &[u8] {
+        &self.pub_key
+    }
+}
+
+impl<Id> Debug for EncapsulationKey<Id>
+where
+    Id: AlgorithmIdentifier,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("PublicKey")
+            .field("algorithm", &self.algorithm)
+            .finish_non_exhaustive()
+    }
+}
+
+/// A set of encrypted bytes produced by [`PublicKey::encapsulate`], and used as an input to [`PrivateKey::decapsulate`].
```

Comment:
Throughout the documentation there are references to "PublicKey" and "PrivateKey", but these are actually referring to `EncapsulationKey` and `DecapsulationKey`, respectively.